pkg/apiserver/handlers.go

```diff
@@ -413,6 +415,65 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
 	return &attribs
 }
 
+func WithActingAs(handler http.Handler, requestContextMapper api.RequestContextMapper, a authorizer.Authorizer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestedSubject := req.Header.Get("Act-As")
+		if len(requestedSubject) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		ctx, exists := requestContextMapper.Get(req)
+		if !exists {
+			forbidden(w, req)
+			return
+		}
+		requestor, exists := api.UserFrom(ctx)
+		if !exists {
+			forbidden(w, req)
+			return
+		}
+
+		actingAsAttributes := &authorizer.AttributesRecord{
+			User:     requestor,
+			Verb:     "act-as",
+			APIGroup: api.GroupName,
+			Resource: "users",
+			// ResourceName:    requestedSubject,
+			ResourceRequest: true,
+		}
+
+		err := a.Authorize(actingAsAttributes)
+		if err != nil {
+			forbidden(w, req)
+			return
+		}
+
+		switch {
+		case strings.HasPrefix(requestedSubject, serviceaccount.ServiceAccountUsernamePrefix):
+			namespace, name, err := serviceaccount.SplitUsername(requestedSubject)
+			if err != nil {
+				forbidden(w, req)
+				return
+			}
+			requestContextMapper.Update(req, api.WithUser(ctx, serviceaccount.UserInfo(namespace, name, "")))
+
+		default:
+			newUser := &user.DefaultInfo{
+				Name: requestedSubject,
+			}
+			requestContextMapper.Update(req, api.WithUser(ctx, newUser))
+		}
+
+		newCtx, _ := requestContextMapper.Get(req)
+		oldUser, _ := api.UserFrom(ctx)
+		newUser, _ := api.UserFrom(newCtx)
+		glog.V(3).Infof("%v is acting as %v on %v", oldUser, newUser, req.URL)
```

Comment:
Is this request going to go into the HTTP log as the original or the impersonated user?  I think the original user is more useful from a debugging and audit logging perspective, so we should try to preserve that.  If it is possible to decorate the logs lines with the impersonated user name, that would be nice, but not as important as logging the original user.